scipy/stats/_bws_test.py

```diff
@@ -0,0 +1,230 @@
+    x, y = np.moveaxis(x, axis, -1), np.moveaxis(y, axis, -1)
+    Ri, Hj = np.sort(x, axis=-1), np.sort(y, axis=-1)
+    n, m = Ri.shape[-1], Hj.shape[-1]
+    i, j = np.arange(1, n+1), np.arange(1, m+1)
+
+    if alternative == 'two-sided':
+        if variant == 'naive':
+            Bx_num = (Ri - (m + n) / n * i) ** 2
+            By_num = (Hj - (m + n) / m * j) ** 2
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * m * (m + n) / n
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * n * (m + n) / m
+        else:
+            Bx_num = (Ri - (m + n + 1) / (n + 1) * i) ** 2
+            By_num = (Hj - (m + n + 1) / (m + 1) * j) ** 2
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * \
+                m * (m + n + 1) / (n + 2)
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * \
+                n * (m + n + 1) / (m + 2)
+
+        Bx = 1 / n * np.sum(Bx_num / Bx_den, axis=-1)
+        By = 1 / m * np.sum(By_num / By_den, axis=-1)
+
+        B = (Bx + By) * 0.5
+    else:
+        if variant == 'naive':
+            Bx_num = (Ri - (m + n) / n * i) * abs(Ri - (m + n) / n * i)
+            By_num = (Hj - (m + n) / m * j) * abs(Hj - (m + n) / m * j)
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * m * (m + n) / n
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * n * (m + n) / m
+        else:
+            Bx_num = (Ri - (m + n + 1) / (n + 1) * i) * \
+                abs(Ri - (m + n + 1) / (n + 1) * i)
+            By_num = (Hj - (m + n + 1) / (m + 1) * j) * \
+                abs(Hj - (m + n + 1) / (m + 1) * j)
+            Bx_den = i / (n + 1) * (1 - i / (n + 1)) * \
+                m * (m + n + 1) / (n + 2)
+            By_den = j / (m + 1) * (1 - j / (m + 1)) * \
+                n * (m + n + 1) / (m + 2)
+
+        Bx = 1 / n * np.sum(Bx_num / Bx_den, axis=-1)
+        By = 1 / m * np.sum(By_num / By_den, axis=-1)
+
+        if alternative == 'greater':
+            B = (Bx - By) * 0.5
+        else:
+            B = (By - Bx) * 0.5
+
+    return B
+
+
+BWSResult = namedtuple('BWSResult', ('statistic', 'pvalue'))
+
+
+def bws_test(x, y, axis=0, alternative="two-sided", variant="naive", n_resamples=9999, random_state=None):
+    r'''Perform the Baumgartner-Weiss-Schindler test on two independent samples.
+
+    The Baumgartner-Weiss-Schindler (BWS) test is a nonparametric test of 
+    the null hypothesis that the distribution underlying sample `x` 
+    is the same as the distribution underlying sample `y`. Unlike 
+    the Kolmogorv-Smirnov, Wilcoxon, and Cramer-Von Mises tests, 
+    the BWS test weights the integral by the variance of the difference
+    in CDFs, emphasizing the tails of the distributions, which increases
+    the power of the test for a lot of applications.
+
+    Parameters
+    ----------
+    x, y : array-like
+        N-d arrays of samples. The arrays must be broadcastable except along
+        the dimension given by `axis`.
+    axis : int, optional
```

Comment:
It's fine to leave `axis` in the code, but I'd recommend removing the public argument *for now*. When we add `_axis_nan_policy` decorator in a follow-up, we'll expose it. The advantage is that you won't have to write tests.